### ISSUE TITLE
ion-datetime Section 508 compliance issues

### ISSUE THREAD

**Issue by koscde** [ionic-team/ionic#15958]
**Ionic version:**  (check one with "x")
(For Ionic 1.x issues, please use https://github.com/ionic-team/ionic-v1)
[ ] **2.x**
[X] **3.x**
[ ] **4.x**

**I'm submitting a ...**  (check one with "x")
[X] bug report
[ ] feature request

**Current behavior:**
Testing at our company and by our client has demonstrated a number of accessibility issues with the ion-datetime component for ionic 3. For any company or Agency that must meet Section 508 compliance requirements, this component is problematic.

**Testing Environment**
•	iOS 12.0.1/VoiceOver screen reader
•	Android 9/TalkBack screen reader
•	Android 9/ Logitech K480 bluetooth keyboard

**Android with TalkBack Enabled**

1. Focus Handling and Visual Indication of Focus
- In Android, with TalkBack running, when the user closes the dialog, focus shifts to the top of the Webview rather than returning to the initiating date field.

- While the user can navigate the dates in the picker using swipe navigation, when the user navigates out of the list of visible items, new items are not displayed (focus indicator/visible text does not track what is announced/programmatically focused; thus, visual focus does not follow the user’s progress through the component). 

- When the picker/spinner is open, the user can navigate into the inactive portion of the screen using swipe navigation (focus is not trapped in the component).

2. Name, Role and State
- When a number is selected, that information is not announced (current state is not indicated).

3. Navigation
- Once a number is selected, the user cannot navigate out of that spinner to the next spinner or to dialog buttons. The user must navigate through the full list before moving to the next series of items (e.g., all the months before navigating to days, all the days before navigating to years). No shortcut is provided for moving between the lists.

**Android with an External Keyboard (Bluetooth Keyboard)**

1. Focus Handling and Visible Indication of Focus
- When the picker/spinner is open, the user can navigate into the inactive portion of the screen using keyboard input (e.g., the tab and arrow keys). Although the background is visually presented as inactive, a keyboard user can navigate into and interact with it (focus is not trapped in the component while active).

- When the user navigates the spinners using Tab key/Shift Tab (to move up and down the list), visual focus is lost after the user leaves the list visible/displayed options (i.e., visual focus does not follow keyboard input). The focus indicator is not visible after the user moves out of the list of currently visible items.

- Navigating by arrow keys appears to have no impact. Moving through a list by up and down arrow keys is an standard keyboard navigation behavior.

- When the user cancels or submits a date and the dialog closes, focus returns to the top of the Web view, rather than going to the date input field.

2. Focus Order
- When user navigates to years, days, or months that are not visible, some options in the list may not be navigable/may be dropped from the list.

3. Navigation
- Once a number is selected, the user cannot navigate out of that spinner to the next spinner or to dialog buttons. The user must navigate through the full list before moving to the next series of items (e.g., all the months before navigating to days). No option is provided for moving between the lists (e.g., up and down arrow to navigate the spinner; left/right arrows to navigate between spinners).

**iOS with VoiceOver Enabled**

1. Focus Handling and Visible Indication of Focus
- Focus does not move into the component when it is opened; if the user cannot see that the dialog is open, there is no indication that anything is displayed (i.e., the presence of the dialog is not announced by the screen reader, and the user must navigate to it).

- The screen reader announces content in each spinner, but only the currently visible dates receive a visual indication of focus. Once the user navigates beyond that abbreviated list, progress is not visually indicated (the spinner does not follow screen reader progress/focus) and no visual indication of focus is provided.

2. Name, Role, State and Value
- If an item is selected, that information is not announced (change in state).

- User selections (day/month/year value, selected by double tapping) are not preserved as the user moves through the UI. When the Done button is double tapped, user selections are not populated in the form.

**WCAG References**

SC 1.3.1: Info and Relationships
http://www.w3.org/WAI/WCAG21/Understanding/info-and-relationships.html 

SC 2.1.1: Keyboard
http://www.w3.org/TR/UNDERSTANDING-WCAG20/keyboard-operation-keyboard-operable.html 

SC 2.4.3: Focus Order
http://www.w3.org/TR/UNDERSTANDING-WCAG20/navigation-mechanisms-focus-order.html

SC 2.4.7: Focus Visible
http://www.w3.org/TR/UNDERSTANDING-WCAG20/navigation-mechanisms-focus-visible.html

SC 4.1.2 Name, Role, Value
http://www.w3.org/WAI/WCAG21/Understanding/name-role-value.html 

**Screencasts**
iOS/VoiceOver: http://www.youtube.com/watch?v=K837CD5YX3I
Android/TalkBack: https://youtu.be/_Yhnji3FG5w
Android/Bluetooth Keyboard: https://youtu.be/wH9uj_3V69k

**Expected behavior:**
The picker should display common UI behaviors for picker components, such as:
* A visual indication of current focus must be provided as the user navigates the component either using a keyboard alone or via a mobile device with screen reader enabled
* The user should be able to determine current location within the component
* The user must be able to navigate to all options within the picker while navigating using a screen reader
* The component must be keyboard operable and respond to keyboard input
* Current state (e.g., "selected") must be programmatically determinable
* The component must preserve user selections as the user navigates the component using a screen reader in either iOS or Android
* When the component is active/open, focus should be trapped in the component until the user closes it
* When the user closes the component, focus should return to the field that caused the datepicker to display

**Steps to reproduce:**
See sample application: http://datetime.squareaxis.com

**Related code:**

**Other information:**
Ionic:

   ionic (Ionic CLI)  : 4.2.1 (/usr/local/lib/node_modules/ionic)
   Ionic Framework    : ionic-angular 3.9.2
   @ionic/app-scripts : 3.1.11

Cordova:

   cordova (Cordova CLI) : 8.1.1 (cordova-lib@8.1.0)
   Cordova Platforms     : android 7.1.1, ios 4.5.5
   Cordova Plugins       : no whitelisted plugins (16 plugins total)

System:

   Android SDK Tools : 26.1.1
   ios-deploy        : 2.0.0
   ios-sim           : 6.1.2
   NodeJS            : v10.5.0 (/usr/local/bin/node)
   npm               : 6.4.1
   OS                : macOS
   Xcode             : Xcode 10.0 Build version 10A255

**Comment by gitdude1458** [comment]
Isn't the easiest solution to use ionic-native datepicker? It will use the platform's date picker, which should meet all the above requirements.

**Comment by koscde** [comment]
Yes, we did try that one using both TalkBack and VoiceOver, external board without screen-reader, and tested focus management/focus trapping both to the picker and back to the originating field when dismissed, but ultimately ran into the same issues.  Our specific need is to have these accessibility concerns resolved with the ion-datetime component.
